rfd/0090-db-mfa-sessions.md

```diff
@@ -0,0 +1,283 @@
+---
+author: Gavin Frazar (gavin.frazar@goteleport.com)
+state: draft
+---
+ 
+# RFD 90 - Database MFA Sessions
+
+## Required Approvers
+- Engineering: Marek `@smallinsky`, Roman `@r0mant`
+- Product: Sasha `@klizhentas`, Xin `@xinding33`
+
+## What
+Use a local proxy tunnel to define the lifetime of a database MFA session, so that database access users are not required to tap an MFA device every minute when per-session-MFA is enabled.
+
+## Why
+Database access via GUI client such as DataGrip requires the user to start a local proxy to connect to: `tsh proxy db --tunnel <db>`.
+
+To support per-session-MFA for the local proxy, we are adding a callback function called on each new connection to the local proxy.
+This callback function will check if the database certs need to be reissued to proxy the connection, and prompt for credentials/MFA if needed.
+
+Currently, database certs are issued with 1 minute ttl if per-session-MFA is enabled as described in [RFD 14](https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md).
+This causes poor UX when the GUI client can establish a new connection at any time,
+and a user may be prompt once per minute for their MFA.
+
+We should remove the 1 minute restriction on database cert TTL to improve UX.
```

Comment:
`tsh proxy db --tunnel` checks cert expiration now and prompts for credentials/mfa as needed to reissue certs - I added that a few weeks ago.
>Instead of removing the 1 minute TTL on the cert, could we instead perform the per-session MFA reissue when you run tsh proxy db --tunnel <db>?

If the cert TTL is 1 minute, then the local proxy will need to refresh its cert after 1 minute as well. Each incoming connection to the local proxy requires the local proxy to establish a new connection to the remote proxy.

Reply:
I see, I naively thought `tsh proxy db` worked the same as `tsh proxy ssh`, which prepares a single proxied connection directly to the node, rather than supporting multiple concurrent connections (I think).

I think it's still worth exploring maintaining the 1 minute TTL if possible, otherwise the security implications of per-session MFA are not consistent across Teleport protocols. If you've already exhausted potentials options and it isn't possible, then I don't intend to block this RFD or anything, I just wanted to make myself available for brainstorming here.

Another potential option could be to change how we enforce per-session MFA. You could perform the per-session MFA reissue and enforcement on `tsh proxy db --tunnel`, which will immediately setup the proxy connection with the 1-minute TTL cert. Then, you could let each db connection through the tunnel use non-mfa certs, and enforce per-session MFA on the db service side by checking that _either_ the connection has per-session MFA certs, or it was proxied from a proxy connection that does have per-session MFA certs.

Reply:
> I see, I naively thought tsh proxy db worked the same as tsh proxy ssh, which prepares a single proxied connection directly to the node, rather than supporting multiple concurrent connections (I think).

Yes, to clarify if it's not clear: `tsh proxy db --tunnel` starts a foreground process that accepts and proxies connections concurrently.

> you could let each db connection through the tunnel use non-mfa certs

Currently the `--tunnel` option allows local connections to connect without any certs at all, since they go through a TLS tunnel setup by the local proxy. The main use-case for this proxy tunnel is to enable GUI clients to connect and the local proxy handles ALPN/SNI TLS. The problem is that GUI clients create new connections all the time, and don't consistently re-use an existing connection. This was especially problematic with MS SQL server.

I think you're suggesting we make the local proxy establish a long-lived TLS connection and multiplex incoming client connections through that? That could work, except that users would need to also disable `disconnect_expired_cert` for DB resources, or the connection would die every minute. I think actually implementing this multiplexing would very difficult though, no? I think we would need a new ALPN wrapper like we did for the "-ping" variant